packages/pds/src/db/message-queue/index.ts

```diff
@@ -0,0 +1,276 @@
+import { RepoStructure } from '@atproto/repo'
+import Database from '..'
+import {
+  AddMember,
+  AddUpvote,
+  CreateNotification,
+  DeleteNotifications,
+  Message,
+  RemoveMember,
+  RemoveUpvote,
+} from './messages'
+import { sql } from 'kysely'
+import { dbLogger as log } from '../../logger'
+import { SceneVotesOnPost } from './tables/sceneVotesOnPost'
+import { AuthStore } from '@atproto/auth'
+import * as repoUtil from '../../util/repo'
+import * as schema from '../../lexicon/schemas'
+import { TID } from '@atproto/common'
+import { MessageQueue } from '../types'
+
+export class SqlMessageQueue implements MessageQueue {
+  constructor(
+    private db: Database,
+    private getAuthStore: (did: string) => AuthStore,
+  ) {}
+
+  async send(messages: Message | Message[]): Promise<void> {
+    const res = await this.db.db
+      .insertInto('message_queue')
+      .values({
+        message: JSON.stringify(messages),
+        read: 0,
+        createdAt: new Date().toISOString(),
+      })
+      .returning('id')
+      .execute()
+    res.forEach((row) => this.process(row.id))
+  }
+
+  async catchup(): Promise<void> {
+    let builder = this.db.db
+      .selectFrom('message_queue')
+      .where('read', '=', 0)
+      .selectAll()
+    if (this.db.dialect !== 'sqlite') {
+      builder = builder.forUpdate()
+    }
+    const res = await builder.execute()
+    await Promise.all(res.map((row) => this.process(row.id)))
```

Comment:
Is there any concern with races if we dont await each `process()` call individually?

Reply:
idt we should rely on the application for serialization. this whole section should be considered async processing. I've tried to write the db transactions in such a way that they have the necessary locks in place to not race